### DISCUSSION TITLE
Binocular Gaze Estimation Method

### DISCUSSION THREAD

**Discussion by yang-ruixin** [pupil-labs/pupil#2301] [Q&A] [answered]
Dear authors,

Thanks for your great work.
In terms of binocular gaze estimation, you cite Lech Swirski: "Gaze estimation on glasses-based stereoscopic displays", Chapter: 7.4.2 Cyclopean gaze estimate.
But in Lech's thesis, it seems there are markers to the back of the eye cameras and there is another camera towards the markers to record the position of the eye camera.
But for Pupil hardware, we don't have the markers and this additional camera.
Where did I miss?

Thanks in advance.

**Comment by papr** [comment, marked as answer]
Hi @yang-ruixin 

[We cite Swirski here](https://github.com/pupil-labs/pupil/blob/master/pupil_src/shared_modules/gaze_mapping/gazer_3d/gazer_headset.py#L263-L286) because we apply their approach of intersecting two monocular gaze rays to determine a cyclopean gaze ray.

We estimate the eye-to-scene-camera poses based [on bundle adjustment](https://github.com/pupil-labs/pupil/blob/81c83a8c79516013179703805ea57f58ca003394/pupil_src/shared_modules/gaze_mapping/gazer_3d/calibrate_3d.py#L34), using the collected data during calibration. Therefore, there is no need for markers or external cameras.

**Reply by yang-ruixin** [reply]
Thank you very much for your quick reply.

Another question, for 3D one eye gaze estimation, I assume we use an eye model and there is no need of calibration at all. 
The input of this eye model is solely from [this 3D pupil detection](https://pye3d-detector.readthedocs.io/en/stable/examples.html).
And the output of this eye model is an one eye gaze estimation in the eye camera. 
(When we want to know the gaze estimation in the world camera, and also how to intersect two monocular gazes, then we need calibraiton.)
Do I understand everything right above?
And would you please tell me where the code of this eye model is?

Thanks in advance again.

**Reply by papr** [reply]
Your understanding is correct. We use https://github.com/pupil-labs/pye3d-detector/ (for its input, we use https://github.com/pupil-labs/pupil-detectors/).

Note that pye3d has a restrictive license. Pupil Labs only allows its usage when
- using it through the Pupil Core software bundles, or
- using it with official Pupil Core hardware or recordings, or
- for research purposes.

**Reply by yang-ruixin** [reply]
Thank you for your reply and your kind reminder for the license issue.
I am doing a part-time research project at university.
Based on pye3d output, [i.e., result_3d](https://pye3d-detector.readthedocs.io/en/latest/examples.html), would you please tell me how to get or draw monocular gaze estimation like [this](https://raw.githubusercontent.com/hysts/pytorch_mpiigaze_demo/master/assets/results/mpiigaze_video00.gif)?
(For their method, in order to draw the gaze, they have a "gaze vector" as input, I am not sure how to get this "gaze vector" based on pye3d output.)

**Reply by papr** [reply]
The `result["circle_3d"]["normal"]` corresponds to the gaze vector in eye camera coordinares. You need to perform a calibration to collect the necessary data to estimate the physical relationship between eye and scene camera, and then being able to [map the vector from eye to scene camera](https://github.com/pupil-labs/pupil/blob/master/pupil_src/shared_modules/gaze_mapping/gazer_3d/gazer_headset.py#L257-L299). `gaze_line_0/1` correspond to the gaze vectors in scene camera coordinates.

**Reply by yang-ruixin** [reply]
Thank you for your reply again.
So not only for gaze estimation in the world camera, even for monocular gaze estimation in the eye camera, we still need calibration?
[For a demo like this](https://raw.githubusercontent.com/hysts/pytorch_mpiigaze_demo/master/assets/results/mpiigaze_video00.gif) (though their camera is far away from eye and pupil eye camera is very close to the eye), I just want to draw a gaze line in the eye camera video like their demo at this moment.

**Reply by papr** [reply]
For gaze estimation in (monocular) eye camera coordinates use `result["circle_3d"]["normal"]`. You do not need calibration for this. Use `result["sphere_center"]` as the origin of the line.

**Reply by yang-ruixin** [reply]
Thank you very much for your kind reply.
